src/fenicsxconcrete/helper.py

```diff
@@ -32,28 +34,202 @@ def to_magnitude(self) -> dict[str, int | str | float]:
+        in a ufl form.
+
+        Args:
+            mesh: A mesh.
+            shape: Local shape of the quadrature space. Example: `shape = 1` for Scalar,
+              `shape = (n, 1)` for vector and `shape = (n,n)` for Tensor.
+        Returns:
+            An array that is equivalent to `quadrature_function.vector.array`.
+        """
+        n_points = self.number_of_points(mesh)
+        n_local = shape if isinstance(shape, int) else shape[0] * shape[1]
+        return np.zeros(n_points * n_local)
+
+
+def _ufl_cell_type_to_basix(cell_type: ufl.Cell) -> basix.CellType:
+    conversion = {
+        ufl.interval: basix.CellType.interval,
+        ufl.triangle: basix.CellType.triangle,
+        ufl.tetrahedron: basix.CellType.tetrahedron,
+        ufl.quadrilateral: basix.CellType.quadrilateral,
+        ufl.hexahedron: basix.CellType.hexahedron,
+    }
+    return conversion[cell_type]
+
+
+class QuadratureEvaluator:
+    """
+    A class that evaluates a ufl expression on a quadrature space.
+
+    Args:
+        ufl_expression: The ufl expression.
+        mesh: The mesh on which we want to evaluate `ufl_expression`
+        rule: The quadrature rule.
+    """
+
+    def __init__(self, ufl_expression: ufl.core.expr.Expr, mesh: df.mesh.Mesh, rule: QuadratureRule) -> None:
+        assert mesh.ufl_cell() == rule.cell_type
+        map_c = mesh.topology.index_map(mesh.topology.dim)
+        self.num_cells = map_c.size_local
+
+        self.cells = np.arange(0, self.num_cells, dtype=np.int32)
+
+        self.expr = df.fem.Expression(ufl_expression, rule.points)
+
+    def evaluate(self, q: np.ndarray | df.fem.Function | None = None) -> np.ndarray | None:
+        """
+        Evaluate the expression.
+
+        Args:
+            q: The object we want to write the result to.
+
+        Returns:
+            A numpy array with the values if `q` is `None`, otherwise the result is written
+            on `q` and `None` is returned.
+        """
+        if q is None:
+            return self.expr.eval(self.cells)
+        elif isinstance(q, np.ndarray):
+            self.expr.eval(self.cells, values=q.reshape(self.num_cells, -1))
+        elif isinstance(q, df.fem.Function):
+            self.expr.eval(self.cells, values=q.vector.array.reshape(self.num_cells, -1))
+            q.x.scatter_forward()
 
 
 def project(
     v: df.fem.Function | ufl.core.expr.Expr, V: df.fem.FunctionSpace, dx: ufl.Measure, u: df.fem.Function | None = None
 ) -> None | df.fem.Function:
     """
     Calculates an approximation of `v` on the space `V`
-
     Args:
         v: The expression that we want to evaluate.
         V: The function space on which we want to evaluate.
         dx: The measure that is used for the integration. This is important, if we want to evaluate
-        a Quadrature function on a normal space.
+        a Quadrature function on a _normal_ space.
```

Comment:
```suggestion
        a Quadrature function on a CG space.
```
?
Not sure what you meant by `_normal_` in this case.

Reply:
Actually, my description is kind of bad. I just meant to say that the measure is important if either `V` is a quadrature space or `v` contains quadrature functions. I mwill correct this